task-definition.tf

```diff
@@ -0,0 +1,107 @@
+locals {
+  default_container_definitions_used = var.container_definitions == "" ? true : false
+
+  default_container_definitions = local.default_container_definitions_used == true ? jsonencode([
+    {
+      name         = var.name
+      image        = "ghcr.io/runatlantis/atlantis:v0.21.0"
+      essential    = true
+      portMappings = [
+        {
+          containerPort = 80
+          hostPort      = 80
+        }
+      ]
+      logConfiguration = {
+        logDriver = "awslogs"
+        options   = {
+          "awslogs-group"         = aws_cloudwatch_log_group.atlantis_container[0].name
+          "awslogs-region"        = "eu-west-1"
+          "awslogs-stream-prefix" = "ecs"
+        }
+      }
+    }
+  ]) : ""
+
+  ecs_task_definition_role_arn = var.ecs_task_definition_role_arn == "" ? aws_iam_role.ecs_task[0].arn : var.ecs_task_definition_role_arn
+}
+
+resource "aws_ecs_task_definition" "atlantis" {
+  family                   = "${var.name}-task"
+  # TODO: Make a separate role for the execution_role_arn
+  execution_role_arn       = local.ecs_task_definition_role_arn
+  task_role_arn            = local.ecs_task_definition_role_arn
+  network_mode             = "awsvpc"
+  requires_compatibilities = ["FARGATE"]
+  cpu                      = var.ecs_task_cpu
+  memory                   = var.ecs_task_memory
+
+  container_definitions = var.container_definitions == "" ? local.default_container_definitions : var.container_definitions
+
+  runtime_platform {
+    operating_system_family = "LINUX"
+    cpu_architecture        = "X86_64"
+  }
+}
+
+data "aws_cloudwatch_log_group" "atlantis_container" {
+  count = var.container_cloudwatch_log_group_name == "" ? 0 : 1
```

Comment:
I agree with you that this makes the code complex and hard to read.
I do think that these conditionals are needed because of different needs that users might have. Some want to use their own ECS cluster or bring their own KMS key into the module.

A way to solve the complexity of this is to put these resources together in a local module, then you only need one condition (count). Which makes it very clear what resources are conditional by grouping them together.

What do you think? Do you have any other suggestions?